### PR TITLE
Bug fix for custom authentication hook, update the status checking of Latex and ImageMagick output

### PR DESCRIPTION
**Pull request by w3shaman**
* Bug fix to allow multiple custom authentication methods work properly. Don't return anything if the custom authentication function returns NULL, so it will continue checking the next available custom authentication methods.
* Improve the way of checking the execution result when converting Latex syntax into image, so the error message on unsuccessful conversion will be displayed more complete and more useful to trace the error.

### CODE REVIEW

File: shared/code/tce_functions_tcecode.php [hunk 1]
```diff
@@ -237,14 +237,14 @@ function F_latex_callback($matches)
             } else {
                 $cmd = 'cd '.K_LATEX_PATH_PICTURE.' && '.K_LATEX_PDFLATEX.' '.$imgpath.'.tex';
                 $sts = exec($cmd, $out, $ret);
-                if (!$sts) {
+                if (!file_exists($imgpath.'.pdf')) {
```
Comment (nicolaasuni) [inline shared/code/tce_functions_tcecode.php:240]: Would the following work for you instead?

```suggestion
                if (!$sts || ($ret != 0)) {
```

File: shared/code/tce_functions_tcecode.php [hunk 2]
```diff
@@ -237,14 +237,14 @@ function F_latex_callback($matches)
...
                     $error = implode("\n", $out);
                 } else {
                     // convert code using ImageMagick
                     $cmd = 'cd '.K_LATEX_PATH_PICTURE.' && '.K_LATEX_PATH_CONVERT.' -density '.(K_LATEX_FORMULA_DENSITY * $dr).' -trim +repage '.$imgpath.'.pdf -depth 8 -quality 100 '.$imgpath.'.'.K_LATEX_IMG_FORMAT;
                     unset($out);
                     $sts = exec($cmd, $out, $ret);
-                    if ($ret != 0) {
+                    if (!file_exists($imgpath.'.'.K_LATEX_IMG_FORMAT)) {
```
Comment (nicolaasuni) [inline shared/code/tce_functions_tcecode.php:247]: Would the following work for you instead?
```suggestion
                   if (($sts === false) || ($ret != 0)) {
```

File: shared/config.default/custom_auth/test.php [hunk 3]
```diff
@@ -0,0 +1,10 @@
+<?php
```
Comment (nicolaasuni) [inline shared/config.default/custom_auth/test.php:1]: There is already a basic example for custom_auth. I'd rather not add more just for testing.

File: shared/custom_auth/test.php [hunk 4]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (nicolaasuni) [inline shared/custom_auth/test.php:1]: There is already a basic example for custom_auth. I'd rather not add more just for testing.

File: shared/code/tce_functions_tcecode.php (hunk 1 above)
Comment (w3shaman) [reply shared/code/tce_functions_tcecode.php:240]: Not working. There is no error in Latex syntax, but all its output in console are displayed. When using file_exists the image is generated and displayed.

File: shared/code/tce_functions_tcecode.php [hunk 5]
```diff
@@ -237,14 +237,14 @@ function F_latex_callback($matches)
             } else {
                 $cmd = 'cd '.K_LATEX_PATH_PICTURE.' && '.K_LATEX_PDFLATEX.' '.$imgpath.'.tex';
                 $sts = exec($cmd, $out, $ret);
-                if (!$sts) {
+                if (!file_exists($imgpath.'.pdf')) {
                     $error = implode("\n", $out);
                 } else {
                     // convert code using ImageMagick
                     $cmd = 'cd '.K_LATEX_PATH_PICTURE.' && '.K_LATEX_PATH_CONVERT.' -density '.(K_LATEX_FORMULA_DENSITY * $dr).' -trim +repage '.$imgpath.'.pdf -depth 8 -quality 100 '.$imgpath.'.'.K_LATEX_IMG_FORMAT;
```
Comment (w3shaman) [inline shared/code/tce_functions_tcecode.php:244]:
```suggestion
                    $cmd = 'cd '.K_LATEX_PATH_PICTURE.' && '.K_LATEX_PATH_CONVERT.' -density '.(K_LATEX_FORMULA_DENSITY * $dr).' -trim +repage '.$imgpath.'.pdf -depth 8 -quality 100 '.$imgpath.'.'.K_LATEX_IMG_FORMAT.' 2>&1';
```

File: shared/code/tce_functions_tcecode.php (hunk 2 above)
Comment (w3shaman) [reply shared/code/tce_functions_tcecode.php:247]: This is working, but I think it's better to also apply this change: [https://github.com/tecnickcom/tcexam/pull/407/files/10546b3e66e38562f4202dfc8c3eb9f8ca76a9fa#r1174262609](url)

File: shared/code/tce_functions_tcecode.php (hunk 1 above)
Comment (nicolaasuni) [reply shared/code/tce_functions_tcecode.php:240]: if $ret=1 there must be something wrong with the conversion.
file_exists has some security issues and I am trying not to use it.
Have you tried testing only for $sts? Would that work for you?

**Comment by nicolaasuni** [comment]
I pushed some changes.
Will you be able to upgrade and check what the actual reported error is and open a new issue?
In the meanwhile I am closing this.

**Pull request closed without merging**